### PR TITLE
doc: V2 Preview documentation updates

### PR DESCRIPTION
**Pull request by edreed**
**What type of PR is this?**

/kind documentation

**What this PR does / why we need it**:

Updates the documentation for the upcoming V2 Preview release.

**Which issue(s) this PR fixes**:

**Requirements**:
- [x] uses [conventional commit messages](https://www.conventionalcommits.org/)
  
- [x] includes documentation
- [ ] adds unit tests
- [ ] tested upgrade from previous version

**Special notes for your reviewer**:

The updates include general formatting issues identified by the Markdown linter.

**Release note**:
```
none
```

### CODE REVIEW

File: docs/design-v2.md [hunk 1]
```diff
@@ -1,133 +1,77 @@
...
-failover latency. It uses shared disks to provision attachment replicas on multiple cluster 
-nodes and integrates with the pod scheduler to ensure a node with a attachment replica is 
-chosen on pod failover.
+Azure Disk CSI Driver V2 improves scalability and reduces pod failover latency. It uses shared disks to provision attachment replicas on multiple cluster nodes and integrates with the pod scheduler to ensure a node with an attachment replica is chosen on pod failover.
 
 ## Architecture and Components
 
-The diagram below shows the components in the Kubernetes control plane (CCP) and cluster 
-nodes, and the Azure services the Mandalay driver uses.
+The diagram below shows the components in the Kubernetes control plane (CCP) and cluster nodes, and the Azure services the Mandalay driver uses.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:7]: @edreed Lets remove all references to Mandalay

File: README.md [hunk 2]
```diff
@@ -1,81 +1,115 @@
...
 ### About
+
 This driver allows Kubernetes to use [Azure Disk](https://azure.microsoft.com/en-us/services/storage/disks/) volume, csi plugin name: `disk.csi.azure.com`
 
-### Project status: GA
+### Project status
+
+V1: GA
+
+V2: Preview
```
Comment (VybavaRamadoss) [inline README.md:15]: Should we say Preview or Alpha?

File: docs/design-v2.md [hunk 3]
```diff
@@ -1,133 +1,77 @@
...
 
 ### Scheduler Extender
 
-The Manadalay scheduler extender influences pod placement by prioritizing healthy nodes where 
-attachment replicas for the required persistent volume(s) already exist (i.e. node(s) to which 
-the managed disk(s) is(are) already attached). It relies on the `AzVolumeAttachment` instances 
-to determine which nodes have attachment replicas, and the heartbeat information in the 
-`AzDriverNode` to determine health. If no attachment replicas for the specified persistent 
-volume currently exist, the scheduler extender will weight all nodes equally.
+The Manadalay scheduler extender influences pod placement by prioritizing healthy nodes where attachment replicas for the required persistent volume(s) already exist (i.e. node(s) to which the managed disk(s) is(are) already attached). It relies on the `AzVolumeAttachment` instances
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:72]: Remove reference to Mandalay

File: docs/design-v2.md (hunk 1 above)
Comment (VybavaRamadoss) [inline docs/design-v2.md:7]: @edreed remove reference to Mandalay

File: docs/design-v2.md [hunk 4]
```diff
@@ -1,133 +1,77 @@
...
+In addition to the CSI Node API Server, this plug-in also provides feedback for pod placement used by the scheduler extender described below.
 
 The node plug-in is deployed on each node in the cluster as a [DaemonSet](https://kubernetes.io/docs/concepts/workloads/controllers/daemonset/).
 
-### Scheduler Extender
+### Scheduler Extender Plug-in
 
-Mandalay provides a [scheduler extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md) 
-that is responsible for influencing pod placements.
+Mandalay provides a [scheduler extender](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/scheduling/scheduler_extender.md) that is responsible for influencing pod placements.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:25]: @edreed remove reference to Mandalay

File: docs/design-v2.md [hunk 5]
```diff
@@ -1,133 +1,77 @@
...
-[ReplicaSet](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/) 
-through [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) 
-with leader election.
+Like the controller plug-in, the scheduler extender is deployed as a [ReplicaSet](https://kubernetes.io/docs/concepts/workloads/controllers/replicaset/) through [Deployment](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/) with leader election.
 
 ## Implementation
 
-This section describes the changes and high-level implementation details of the different 
-components in Mandalay.
+This section describes the changes and high-level implementation details of the different components in Mandalay.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:31]: @edreed Remove reference to Mandalay

File: docs/design-v2.md [hunk 6]
```diff
@@ -1,133 +1,77 @@
...
 
 ### StorageClass
 
-In addition to the existing StorageClass parameters described in 
-[driver-parameters](driver-parameters.md), Mandalay supports the following:
+In addition to the existing StorageClass parameters described in [driver-parameters](driver-parameters.md), Mandalay supports the following:
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:35]: @edreed Remove reference to Mandalay

File: docs/design-v2.md [hunk 7]
```diff
@@ -1,133 +1,77 @@
...
+| `maxShares` | The total number of shared disk mounts allowed for the disk. Setting the value to 2 or more enables attachment replicas. | Supported values depend on the disk size. See [Share an Azure managed disk](https://docs.microsoft.com/en-us/azure/virtual-machines/disks-shared) for supported values. | No | 1 |
+| `maxMountReplicaCount` | The number of replicas attachments to maintain. | This value must be in the range `[0..(maxShares - 1)]` | No | If `accessMode` is `ReadWriteMany`, the default is `0`. Otherwise, the default is `maxShares - 1` |
 
 ### Custom Resources and Controllers
 
-Mandalay uses 3 different custom resources defined in the `azure-disk-csi` namespace to 
-orchestrate disk management and facilitate pod failover to nodes with attachment replica. A 
-controller for each resource watches for and responds to changes to its custom resource 
-instances.
+Mandalay uses 3 different custom resources defined in the `azure-disk-csi` namespace to orchestrate disk management and facilitate pod failover to nodes with attachment replica. A controller for each resource watches for and responds to changes to its custom resource instances.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:44]: @edreed Remove reference to Mandalay

File: docs/design-v2.md (hunk 3 above)
Comment (VybavaRamadoss) [inline docs/design-v2.md:72]: @edreed Remove reference to Manadalay

File: README.md [hunk 8]
```diff
@@ -1,81 +1,115 @@
...
+
+- The driver depends on [cloud provider config file](https://kubernetes-sigs.github.io/cloud-provider-azure/install/configs/#setting-azure-cloud-provider-from-kubernetes-secrets), usually it's `/etc/kubernetes/azure.json` on all kubernetes nodes deployed by [AKS](https://docs.microsoft.com/en-us/azure/aks/) or [aks-engine](https://github.com/Azure/aks-engine), here is [azure.json example](./deploy/example/azure.json).
+  > To specify a different cloud provider config file, create `azure-cred-file` configmap before driver installation, e.g. for OpenShift, it's `/etc/kubernetes/cloud.conf` (make sure config file path is in the `volumeMounts.mountPath`)
+  >
+  > ```console
+  > kubectl create configmap azure-cred-file --from-literal=path="/etc/kubernetes/cloud.conf" --from-literal=path-windows="C:\\k\\cloud.conf" -n kube-system
+  > ```
+  >
+- This driver also supports [read cloud config from kuberenetes secret](./docs/read-from-secret.md).
+- If cluster identity is [Managed Service Identity(MSI)](https://docs.microsoft.com/en-us/azure/aks/use-managed-identity), make sure user assigned identity has `Contributor` role on node resource group
```
Comment (VybavaRamadoss) [inline README.md:54]: @edreed Should we add some additional information here e.g. to troubleshoot if the user runs into permission issues?

File: docs/design-v2.md [hunk 9]
```diff
@@ -1,133 +1,77 @@
...
-chosen on pod failover.
+Azure Disk CSI Driver V2 improves scalability and reduces pod failover latency. It uses shared disks to provision attachment replicas on multiple cluster nodes and integrates with the pod scheduler to ensure a node with an attachment replica is chosen on pod failover.
 
 ## Architecture and Components
 
-The diagram below shows the components in the Kubernetes control plane (CCP) and cluster 
-nodes, and the Azure services the Mandalay driver uses.
+The diagram below shows the components in the Kubernetes control plane (CCP) and cluster nodes, and the Azure services the Mandalay driver uses.
 
 ![Mandalay Architecture](images/mandalay_arch.png)
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:9]: @edreed Remove reference to Mandalay

File: docs/design-v2.md [hunk 10]
```diff
@@ -1,133 +1,77 @@
...
-controller for each resource watches for and responds to changes to its custom resource 
-instances.
+Mandalay uses 3 different custom resources defined in the `azure-disk-csi` namespace to orchestrate disk management and facilitate pod failover to nodes with attachment replica. A controller for each resource watches for and responds to changes to its custom resource instances.
 
 #### `AzDriverNode` Resource and Controller
 
-The `AzDriverNode` custom resource represents a node in the cluster where the Mandalay node 
-plug-in runs. An instance of `AzDriverNode` is created when the node plug-in starts. The node 
-plug-in periodically updates the heartbeat in the `Status` field.
+The `AzDriverNode` custom resource represents a node in the cluster where the Mandalay node plug-in runs. An instance of `AzDriverNode` is created when the node plug-in starts. The node plug-in periodically updates the heartbeat in the `Status` field.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:48]: @edreed Remove reference to Mandalay

File: docs/design-v2.md [hunk 11]
```diff
@@ -1,133 +1,77 @@
...
+to determine which nodes have attachment replicas, and the heartbeat information in the `AzDriverNode` to determine health. If no attachment replicas for the specified persistent volume currently exist, the scheduler extender will weight all nodes equally.
 
 ### Provisioner Library
 
-The Provisioner Library is a common library to abstract the underlying platform for all 
-Mandalay plugins, services and controllers. It handles the platform-specific details of 
-performing volume operations such as (but not necessarily limited to) create, delete, attach, 
-detach, snapshot, stage, unstage, mount, unmount, etc.
+The Provisioner Library is a common library to abstract the underlying platform for all Mandalay plugins, services and controllers. It handles the platform-specific details of performing volume operations such as (but not necessarily limited to) create, delete, attach, detach, snapshot, stage, unstage, mount, unmount, etc.
```
Comment (VybavaRamadoss) [inline docs/design-v2.md:77]: @edreed Remove reference to Mandalay

**Review by VybavaRamadoss — COMMENTED** [review]
Doc Page Review

File: README.md (hunk 2 above)
Comment (edreed) [reply README.md:15]: I believe Preview is fine here. I am using the Kubernetes convention for labeling. `alpha` just means the API is not stable and may change. In our case, this references the CRDs but could be features as well.

File: docs/design-v2.md (hunk 1 above)
Comment (edreed) [reply docs/design-v2.md:7]: Agreed.

File: README.md (hunk 8 above)
Comment (edreed) [reply README.md:54]: Done

File: docs/design-v2.md [hunk 12]
```diff
@@ -1,133 +1,77 @@
...
+The `AzVolume` custom resource represents the managed disk of a `PersistentVolume`. The controller for `AzVolume` runs in the controller plug-in. It watches for and reconciles changes in the `AzVolume` instances.
 
-An `AzVolume` instance is created by the `CreateVolume` API to the CSI Controller plug-in. The 
-`AzVolume` controller responds to the new instance by creating a managed disk using the 
-parameters in the referenced `StorageClass`. When disk creation is complete, the controller 
-sets the `.Status.State` field to `Created` or suitable error state on failure. It also adds a 
-finalizer if creation was successful. The `CreateVolume` request completes when the `AzVolume` 
-state is updated.
+An `AzVolume` instance is created by the `CreateVolume` API to the CSI Controller plug-in. The `AzVolume` controller responds to the new instance by creating a managed disk using the parameters in the referenced `StorageClass`. When disk creation is complete, the controller sets the `.Status.State` field to `Created` or suitable error state on failure. It also adds a finalizer if creation was successful. The `CreateVolume` request completes when the `AzVolume` state is updated.
```
Comment (sunpa93) [inline docs/design-v2.md:56]: >It also adds a finalizer if creation was successful.
It adds a finalizer to the instance upon CRI creation.

File: docs/design-v2.md [hunk 13]
```diff
@@ -1,133 +1,77 @@
...
+
+An `AzVolumeAttachment` instance representing the primary node attachment is created by the `ControllerPublishVolume` API in the CSI Controller plug-in. If an instance for the current node already exists, it is updated to represent the primary node attachment. The `AzVolumeAttachment` controller then attaches the manage disk to the primary node (downgrading to attachment replica or detaching any previous primary attachment if one exists). It then creates additional `AzVolumeAttachment` instances representing attachment replicas and attaches the shared manage disk to a number of backup nodes as specified by the `maxMountReplicaCount` parameter in the `StorageClass` instance of the `PersistentVolumeClaim` of the managed disk. The `ControllerPublishVolume` request is complete once the `AzVolumeAttachment` instance for primary node has been crerated. As each the attachment request completes, the controller sets the `.Status.State` field to `Attached` and adds a finalizer to each `AzVolumeAttachment` instance. When the `NodeStageVolume` API is called, the CSI Node plug-in will wait for the its
```
Comment (sunpa93) [inline docs/design-v2.md:65]: > (downgrading to attachment replica or detaching any previous primary attachment if one exists).
downgrade?

Comment (sunpa93) [reply docs/design-v2.md:65]: sp: The `ControllerPublishVolume` request is complete once the `AzVolumeAttachment` instance for primary node has been ~crerated~ created

Comment (sunpa93) [reply docs/design-v2.md:65]: > the controller sets the `.Status.State` field to `Attached` and adds a finalizer to each `AzVolumeAttachment` instance

This is misleading as finalizers are added as soon as the CR instance is created.

File: docs/design-v2.md [hunk 14]
```diff
@@ -1,133 +1,77 @@
...
+`AzVolumeAttachment` instance's state to reach `Attached` before staging the mount point to the disk.
+
+When the `ControllerUnpublishVolume` API in the CSI Controller plug-in is called, it deletes the `AzVolumeAttachment` instance for the primary node. The controller responds by detaching the managed disk from the primary node. It removes the finalizer from the `AzVolumeAttachment` instance when the detach operation completes. The `ControllerUnpublishVolume` request is complete when the detach operation has completed and corresponding `AzVolumeAttachment` has been removed from the object store.
```
Comment (sunpa93) [inline docs/design-v2.md:68]: And the controller schedules a garbage collection to delete replica attachments if no subsequent `ControllerPublishVolume` request is made for the volume after its detachment from the primary node within 5 minutes.

File: docs/design-v2.md (hunk 12 above)
Comment (edreed) [reply docs/design-v2.md:56]: I will drop mention of the finalizers. I don't believe when they are added is critical to understanding the overall design and workflows.

File: docs/design-v2.md (hunk 13 above)
Comment (edreed) [reply docs/design-v2.md:65]: Fixed.

File: docs/design-v2.md (hunk 14 above)
Comment (edreed) [reply docs/design-v2.md:68]: Done.

**Pull request merged**